**Importing libraries**

In [1]:
#Importing necessary libraries
import pandas as pd
import numpy as np

# Matplot
import matplotlib.pyplot as plt
# Seaborn
import seaborn as sns

# Cell-to-cell script
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

import scipy.stats as st

**Getting the data**

In [2]:
# The datacomes from the Taarifa waterpoints dashboard.
#Taarifa aggregates data from the Tanzania Ministry of Water. I download this data throught 'Drive Data' data resources.
#The data is already splitted: feature_data,label_data. I will merge the two for the  initial exploration with the data.

In [3]:
#Get the data
df_features = pd.read_csv('data_features.csv') 
df_label=pd.read_csv('data_labels.csv') 

In [4]:
df_features.head()
df_features.shape

id  amount_tsh date_recorded        funder  gps_height     installer  \
0  69572      6000.0    2011-03-14         Roman        1390         Roman   
1   8776         0.0    2013-03-06       Grumeti        1399       GRUMETI   
2  34310        25.0    2013-02-25  Lottery Club         686  World vision   
3  67743         0.0    2013-01-28        Unicef         263        UNICEF   
4  19728         0.0    2011-07-13   Action In A           0       Artisan   

   longitude   latitude              wpt_name  num_private  ... payment_type  \
0  34.938093  -9.856322                  none            0  ...     annually   
1  34.698766  -2.147466              Zahanati            0  ...    never pay   
2  37.460664  -3.821329           Kwa Mahundi            0  ...   per bucket   
3  38.486161 -11.155298  Zahanati Ya Nanyumbu            0  ...    never pay   
4  31.130847  -1.825359               Shuleni            0  ...    never pay   

  water_quality quality_group      quantity  quantity_group  \
0          soft          good        enough          enough   
1          soft          good  insufficient    insufficient   
2          soft          good        enough          enough   
3          soft          good           dry             dry   
4          soft          good      seasonal        seasonal   

                 source           source_type  source_class  \
0                spring                spring   groundwater   
1  rainwater harvesting  rainwater harvesting       surface   
2                   dam                   dam       surface   
3           machine dbh              borehole   groundwater   
4  rainwater harvesting  rainwater harvesting       surface   

               waterpoint_type waterpoint_type_group  
0           communal standpipe    communal standpipe  
1           communal standpipe    communal standpipe  
2  communal standpipe multiple    communal standpipe  
3  communal standpipe multiple    communal standpipe  
4           communal standpipe    communal standpipe  

[5 rows x 40 columns]

(59400, 40)

In [5]:
df_label.head()
df_label.shape

id    status_group
0  69572      functional
1   8776      functional
2  34310      functional
3  67743  non functional
4  19728      functional

(59400, 2)

In [6]:
#Both data set has the same number of rows 59400. I will merge both to get my initial data set.

In [7]:
df = pd.merge(df_features,df_label,on='id')
df.shape
df.head()

(59400, 41)

id  amount_tsh date_recorded        funder  gps_height     installer  \
0  69572      6000.0    2011-03-14         Roman        1390         Roman   
1   8776         0.0    2013-03-06       Grumeti        1399       GRUMETI   
2  34310        25.0    2013-02-25  Lottery Club         686  World vision   
3  67743         0.0    2013-01-28        Unicef         263        UNICEF   
4  19728         0.0    2011-07-13   Action In A           0       Artisan   

   longitude   latitude              wpt_name  num_private  ... water_quality  \
0  34.938093  -9.856322                  none            0  ...          soft   
1  34.698766  -2.147466              Zahanati            0  ...          soft   
2  37.460664  -3.821329           Kwa Mahundi            0  ...          soft   
3  38.486161 -11.155298  Zahanati Ya Nanyumbu            0  ...          soft   
4  31.130847  -1.825359               Shuleni            0  ...          soft   

  quality_group      quantity  quantity_group                source  \
0          good        enough          enough                spring   
1          good  insufficient    insufficient  rainwater harvesting   
2          good        enough          enough                   dam   
3          good           dry             dry           machine dbh   
4          good      seasonal        seasonal  rainwater harvesting   

            source_type source_class              waterpoint_type  \
0                spring  groundwater           communal standpipe   
1  rainwater harvesting      surface           communal standpipe   
2                   dam      surface  communal standpipe multiple   
3              borehole  groundwater  communal standpipe multiple   
4  rainwater harvesting      surface           communal standpipe   

  waterpoint_type_group    status_group  
0    communal standpipe      functional  
1    communal standpipe      functional  
2    communal standpipe      functional  
3    communal standpipe  non functional  
4    communal standpipe      functional  

[5 rows x 41 columns]

In [8]:
#By default not all the columns are display. We want to see them all 
pd.set_option('max_columns', None)
df.head(50)

id  amount_tsh date_recorded                    funder  gps_height  \
0   69572      6000.0    2011-03-14                     Roman        1390   
1    8776         0.0    2013-03-06                   Grumeti        1399   
2   34310        25.0    2013-02-25              Lottery Club         686   
3   67743         0.0    2013-01-28                    Unicef         263   
4   19728         0.0    2011-07-13               Action In A           0   
5    9944        20.0    2011-03-13       Mkinga Distric Coun           0   
6   19816         0.0    2012-10-01                      Dwsp           0   
7   54551         0.0    2012-10-09                     Rwssp           0   
8   53934         0.0    2012-11-03                  Wateraid           0   
9   46144         0.0    2011-08-03               Isingiro Ho           0   
10  49056         0.0    2011-02-20                   Private          62   
11  50409       200.0    2013-02-18                    Danida        1062   
12  36957         0.0    2012-10-14              World Vision           0   
13  50495         0.0    2013-03-15   Lawatefuka Water Supply        1368   
14  53752         0.0    2012-10-20                     Biore           0   
15  61848         0.0    2011-08-04                     Rudep        1645   
16  48451       500.0    2011-07-04                    Unicef        1703   
17  58155         0.0    2011-09-04                    Unicef        1656   
18  34169         0.0    2011-07-22                    Hesawa        1162   
19  18274       500.0    2011-02-22                    Danida        1763   
20  48375       200.0    2011-02-27                       Twe        2216   
21   6091         0.0    2013-02-10                      Dwsp           0   
22  58500         0.0    2011-10-04                    Unicef        1510   
23  37862       500.0    2013-11-03                       Isf         672   
24  51058         0.0    2013-01-21  African Development Bank        1645   
25  22308         0.0    2013-02-25    Government Of Tanzania        1273   
26  55012       500.0    2013-01-16                    Sobodo         200   
27  20145         0.0    2011-07-11                    Hesawa           0   
28  19685         0.0    2013-03-05    Government Of Tanzania        1443   
29  69124         0.0    2013-03-16   Lawatefuka Water Supply        1256   
30  46804         0.0    2011-03-23                     Water           0   
31   6696        50.0    2011-03-16        Private Individual         180   
32  57238         0.0    2013-03-19    Government Of Tanzania        1481   
33  12402         0.0    2011-03-11                      Undp        1638   
34  41583         0.0    2011-02-23                       NaN         -41   
35  57355         0.0    2013-03-28                 Not Known        1546   
36  67359         0.0    2011-07-16                    Danida           0   
37  60048         0.0    2011-03-27                     Kirde           0   
38  16583      4000.0    2011-02-20                      Cefa        1955   
39     25         0.0    2013-02-11    Government Of Tanzania        1348   
40  70238        25.0    2013-10-03                 Ces(gmbh)         710   
41  12796       500.0    2011-03-12    Government Of Tanzania        2469   
42  52019       500.0    2011-07-07            European Union        1803   
43  19282         0.0    2013-01-15                       NaN        1642   
44   1843        20.0    2013-03-18    Government Of Tanzania        1007   
45  12958         0.0    2012-10-22              World Vision           0   
46  45111        20.0    2013-02-05                       Lga         240   
47  13620         0.0    2011-07-27                       NaN           0   
48  18864      4000.0    2011-02-20                      Cefa        1982   
49  25182      1500.0    2011-04-04          District Council        2169   

               installer  longitude      latitude                wpt_name  \
0 

**DATA EXPLORATION: Initial approach to the data**

In [9]:
df.shape

(59400, 41)

In [10]:
#The data set provides information about the waterpoints in Tanzania. It includes 59400 total records and 42 columns.
#Let´s analyses the  meaning of each column

In [11]:
df.columns

Index(['id', 'amount_tsh', 'date_recorded', 'funder', 'gps_height',
       'installer', 'longitude', 'latitude', 'wpt_name', 'num_private',
       'basin', 'subvillage', 'region', 'region_code', 'district_code', 'lga',
       'ward', 'population', 'public_meeting', 'recorded_by',
       'scheme_management', 'scheme_name', 'permit', 'construction_year',
       'extraction_type', 'extraction_type_group', 'extraction_type_class',
       'management', 'management_group', 'payment', 'payment_type',
       'water_quality', 'quality_group', 'quantity', 'quantity_group',
       'source', 'source_type', 'source_class', 'waterpoint_type',
       'waterpoint_type_group', 'status_group'],
      dtype='object')

**VARIABLE EXPLANATION:**

**amount_tsh** - Total static head (amount water available to waterpoint)

**date_recorded** - The date the row was entered

**funder** - Who funded the well

**gps_height** - Altitude of the well

**installer** - Organization that installed the well

**longitude** - GPS coordinate

**latitude** - GPS coordinate

**wpt_name**- Name of the waterpoint if there is one

**num_private** -

**basin** - Geographic water basin

**subvillage** - Geographic location

**region** - Geographic location

**region_code** - Geographic location (coded)

**district_code** - Geographic location (coded)

**lga** - Geographic location

**ward** - Geographic location

**population** - Population around the well

**public_meeting** - True/False

**recorded_by** - Group entering this row of data

**scheme_management** - Who operates the waterpoint

**scheme_name** - Who operates the waterpoint

**permit** - If the waterpoint is permitted

**construction_year** - Year the waterpoint was constructed

**extraction_type** - The kind of extraction the waterpoint uses

**extraction_type_group** - The kind of extraction the waterpoint uses

**extraction_type_class** - The kind of extraction the waterpoint uses

**management** - How the waterpoint is managed

**management_group** - How the waterpoint is managed

**payment** - What the water costs

**payment_type** - What the water costs

**water_quality** - The quality of the water

**quality_group** - The quality of the water

**quantity** - The quantity of water

**quantity_group** - The quantity of water

**source** - The source of the water

**source_type** - The source of the water

**source_class** - The source of the water

**waterpoint_type** - The kind of waterpoint

**waterpoint_type_group** - The kind of waterpoint

**status_group** - If the waterpoint is operational, needs repair or not operational (This will be our label for the project)


In [12]:
#Let´s have overview of our data set
df.info()
# We have  a total of 59400 entries and 41 columns. 
# We have some null values in several columns. We will have to lead with them.
# Data types: int64(7),float64(3) and object(31). Most of our dataset is categorical data. We find some that is datatime.

<class 'pandas.core.frame.DataFrame'>
Int64Index: 59400 entries, 0 to 59399
Data columns (total 41 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     59400 non-null  int64  
 1   amount_tsh             59400 non-null  float64
 2   date_recorded          59400 non-null  object 
 3   funder                 55765 non-null  object 
 4   gps_height             59400 non-null  int64  
 5   installer              55745 non-null  object 
 6   longitude              59400 non-null  float64
 7   latitude               59400 non-null  float64
 8   wpt_name               59400 non-null  object 
 9   num_private            59400 non-null  int64  
 10  basin                  59400 non-null  object 
 11  subvillage             59029 non-null  object 
 12  region                 59400 non-null  object 
 13  region_code            59400 non-null  int64  
 14  district_code          59400 non-null  int64  
 15  lg

In [13]:
#Split numerical and categorical values
df_num = df.select_dtypes(np.number)
df_cat = df.select_dtypes(np.object)

### NUMERICAL VARIABLES

In [14]:
df_num

id  amount_tsh  gps_height  longitude   latitude  num_private  \
0      69572      6000.0        1390  34.938093  -9.856322            0   
1       8776         0.0        1399  34.698766  -2.147466            0   
2      34310        25.0         686  37.460664  -3.821329            0   
3      67743         0.0         263  38.486161 -11.155298            0   
4      19728         0.0           0  31.130847  -1.825359            0   
...      ...         ...         ...        ...        ...          ...   
59395  60739        10.0        1210  37.169807  -3.253847            0   
59396  27263      4700.0        1212  35.249991  -9.070629            0   
59397  37057         0.0           0  34.017087  -8.750434            0   
59398  31282         0.0           0  35.861315  -6.378573            0   
59399  26348         0.0         191  38.104048  -6.747464            0   

       region_code  district_code  population  construction_year  
0               11              5         109               1999  
1               20              2         280               2010  
2               21              4         250               2009  
3               90             63          58               1986  
4               18              1           0                  0  
...            ...            ...         ...                ...  
59395            3              5         125               1999  
59396           11              4          56               1996  
59397           12              7           0                  0  
59398            1              4           0                  0  
59399            5              2         150               2002  

[59400 rows x 10 columns]

In [15]:
#checking NaN values
df_num.isnull().sum()

id                   0
amount_tsh           0
gps_height           0
longitude            0
latitude             0
num_private          0
region_code          0
district_code        0
population           0
construction_year    0
dtype: int64

In [16]:
#There is NaN values in the numerical variables

In [17]:
#Lets checks the value for each columns.It seems that some columns has a lot of 0 values. Are they null values?i.e:num_private

In [18]:
#checking unique values and the exact value and information for each one.
df_num.nunique()
for x in df_num:
    df_num[x].value_counts()

id                   59400
amount_tsh              98
gps_height            2428
longitude            57516
latitude             57517
num_private             65
region_code             27
district_code           20
population            1049
construction_year       55
dtype: int64

0        1
19811    1
38200    1
34106    1
36155    1
        ..
64138    1
62091    1
51852    1
49805    1
2047     1
Name: id, Length: 59400, dtype: int64

0.0         41639
500.0        3102
50.0         2472
1000.0       1488
20.0         1463
            ...  
250000.0        1
220.0           1
350000.0        1
138000.0        1
38000.0         1
Name: amount_tsh, Length: 98, dtype: int64

 0       20438
-15         60
-16         55
-13         55
 1290       52
         ...  
 2332        1
 2202        1
 2236        1
-57          1
 2038        1
Name: gps_height, Length: 2428, dtype: int64

0.000000     1812
32.977191       2
32.919861       2
37.542785       2
39.105307       2
             ... 
37.141893       1
35.839879       1
33.638301       1
35.347266       1
35.164770       1
Name: longitude, Length: 57516, dtype: int64

-2.000000e-08    1812
-2.494546e+00       2
-6.983183e+00       2
-7.056923e+00       2
-7.056372e+00       2
                 ... 
-6.162973e+00       1
-1.595138e+00       1
-2.007878e+00       1
-6.787691e+00       1
-1.077290e+01       1
Name: latitude, Length: 57517, dtype: int64

0      58643
6         81
1         73
8         46
5         46
       ...  
42         1
136        1
35         1
131        1
94         1
Name: num_private, Length: 65, dtype: int64

11    5300
17    5011
12    4639
3     4379
5     4040
18    3324
19    3047
2     3024
16    2816
10    2640
4     2513
1     2201
13    2093
14    1979
20    1969
15    1808
6     1609
21    1583
80    1238
60    1025
90     917
7      805
99     423
9      390
24     326
8      300
40       1
Name: region_code, dtype: int64

1     12203
2     11173
3      9998
4      8999
5      4356
6      4074
7      3343
8      1043
30      995
33      874
53      745
43      505
13      391
23      293
63      195
62      109
60       63
0        23
80       12
67        6
Name: district_code, dtype: int64

0       21381
1        7025
200      1940
150      1892
250      1681
        ...  
1160        1
3241        1
1960        1
2248        1
1439        1
Name: population, Length: 1049, dtype: int64

0       20709
2010     2645
2008     2613
2009     2533
2000     2091
2007     1587
2006     1471
2003     1286
2011     1256
2004     1123
2012     1084
2002     1075
1978     1037
1995     1014
2005     1011
1999      979
1998      966
1990      954
1985      945
1980      811
1996      811
1984      779
1982      744
1994      738
1972      708
1974      676
1997      644
1992      640
1993      608
2001      540
1988      521
1983      488
1975      437
1986      434
1976      414
1970      411
1991      324
1989      316
1987      302
1981      238
1977      202
1979      192
1973      184
2013      176
1971      145
1960      102
1967       88
1963       85
1968       77
1969       59
1964       40
1962       30
1961       21
1965       19
1966       17
Name: construction_year, dtype: int64

In [19]:
#We find that there are several columns with important amount of '0' values. Let´s check what this means. 
#Can we consider some of them as  NaN values?

In [20]:
#let´s check the total of values=0 in each row
df_num.apply(lambda column:(column == 0).sum())
df_num.apply(lambda column:(column == 0).sum())/len(df_num)

id                       1
amount_tsh           41639
gps_height           20438
longitude             1812
latitude                 0
num_private          58643
region_code              0
district_code           23
population           21381
construction_year    20709
dtype: int64

id                   0.000017
amount_tsh           0.700993
gps_height           0.344074
longitude            0.030505
latitude             0.000000
num_private          0.987256
region_code          0.000000
district_code        0.000387
population           0.359949
construction_year    0.348636
dtype: float64

In [21]:
#num_private has almost all their values = o. We delete this column because it does not give us any important information.

In [22]:
df_num = df_num.drop(['num_private'], axis=1)

In [23]:
#amount_tsh has 70% of values = 0. What does this means? 
#Are they correct values so we have to keep this columns or are all missing data so  we should delete this column?
#Let´s analyze this variable

In [24]:
#This featureis related to the water available to waterpoint.Static head measures the total vertical distance that a pump raises water.
#The data shows that although most of the values are 0 the well is still functional.This is hard to understand, how a well could be functional with this value as 0.
#For this reaseon, we consider this zero values as NaN ( error in the collection of the data).This errors are the 70% of the total.
#We delete this column from our data. 

df_num = df_num.drop(['amount_tsh'], axis=1)

# If I had time I should explore if the zeros are related to the type of waterpoint.

In [25]:
#There are some features with big amount of 0 values. How I should deal with them??
# gps_height           0.344074 -- is geograpical feature. Should  I use it in our model? DROP IT
# longitude            0.030505 --is geograpical feature. Should  I use it in our model? DROP IT
# latitude             0.000000 --is geograpical feature. Should  I use it in our model? DROP IT
# population           0.359949 -- I think this is important.
# there are 7025 values = 1 It makes no sense!!I should have treat values = 1 as null. Use  KNN or media to replace.
# construction_year    0.348636 -- I think this is important.Replace with the mode.

#REGION and DISTRICT code. There are numerical but maybe should be categorical. Should I include them in our model? DROP

In [26]:
# CONSTRUCTION YEAR: we will creat a new column using this one and the actual year(2021)  to have how old is the well (total yeares since construction).
df_num['years_since_construction'] =  np.where(df_num['construction_year']!= 0 , 2021-(df_num['construction_year']),0)
df_num[['years_since_construction']]
# WHERE FUNCTION; first the condiction, second result for fitting the condition (TRUE), last result if not fitting the condiction(FALSE)
# This function only works in this case: CONDITION, TRUE STATMENT, FALSE STATMENT

years_since_construction
0                            22
1                            11
2                            12
3                            35
4                             0
...                         ...
59395                        22
59396                        25
59397                         0
59398                         0
59399                        19

[59400 rows x 1 columns]

In [27]:
df_num.columns

Index(['id', 'gps_height', 'longitude', 'latitude', 'region_code',
       'district_code', 'population', 'construction_year',
       'years_since_construction'],
      dtype='object')

In [28]:
#FINAL NUMERICAL DATA SET: Selecting our final df_num according to the explained above:
df_num = df_num.drop(['id', 'gps_height', 'longitude', 'latitude', 'region_code',
       'district_code','construction_year'], axis=1)
df_num

population  years_since_construction
0             109                        22
1             280                        11
2             250                        12
3              58                        35
4               0                         0
...           ...                       ...
59395         125                        22
59396          56                        25
59397           0                         0
59398           0                         0
59399         150                        19

[59400 rows x 2 columns]

In [29]:
#DUDAAAAAS!!!
#POPULATION
# substituting with the MEAN 
# df_num['population'].hist()

#we use histogram to check if this is consistent ( do not alter my data very strong)
# mean_population= (df_num['population']>1).mean()
mean_population= (df_num['population']).mean()
mean_population
# df_num=df_num.replace({'population': {0: mean_population}}) 
# print (data)

179.90998316498317

### CATEGORICAL VARIABLES

In [30]:
df_cat.head()
df_cat.shape

date_recorded        funder     installer              wpt_name  \
0    2011-03-14         Roman         Roman                  none   
1    2013-03-06       Grumeti       GRUMETI              Zahanati   
2    2013-02-25  Lottery Club  World vision           Kwa Mahundi   
3    2013-01-28        Unicef        UNICEF  Zahanati Ya Nanyumbu   
4    2011-07-13   Action In A       Artisan               Shuleni   

                     basin  subvillage   region        lga        ward  \
0               Lake Nyasa    Mnyusi B   Iringa     Ludewa    Mundindi   
1            Lake Victoria     Nyamara     Mara  Serengeti       Natta   
2                  Pangani     Majengo  Manyara  Simanjiro     Ngorika   
3  Ruvuma / Southern Coast  Mahakamani   Mtwara   Nanyumbu    Nanyumbu   
4            Lake Victoria  Kyanyamisa   Kagera    Karagwe  Nyakasimbi   

  public_meeting              recorded_by scheme_management  \
0           True  GeoData Consultants Ltd               VWC   
1            NaN  GeoData Consultants Ltd             Other   
2           True  GeoData Consultants Ltd               VWC   
3           True  GeoData Consultants Ltd               VWC   
4           True  GeoData Consultants Ltd               NaN   

                   scheme_name permit extraction_type extraction_type_group  \
0                        Roman  False         gravity               gravity   
1                          NaN   True         gravity               gravity   
2  Nyumba ya mungu pipe scheme   True         gravity               gravity   
3                          NaN   True     submersible           submersible   
4                          NaN   True         gravity               gravity   

  extraction_type_class management management_group         payment  \
0               gravity        vwc       user-group    pay annually   
1               gravity        wug       user-group       never pay   
2               gravity        vwc       user-group  pay per bucket   
3           submersible        vwc       user-group       never pay   
4               gravity      other            other       never pay   

  payment_type water_quality quality_group      quantity quantity_group  \
0     annually          soft          good        enough         enough   
1    never pay          soft          good  insufficient   insufficient   
2   per bucket          soft          good        enough         enough   
3    never pay          soft          good           dry            dry   
4    never pay          soft          good      seasonal       seasonal   

                 source           source_type source_class  \
0                spring                spring  groundwater   
1  rainwater harvesting  rainwater harvesting      surface   
2                   dam                   dam      surface   
3           machine dbh              borehole  groundwater   
4  rainwater harvesting  rainwater harvesting      surface   

               waterpoint_type waterpoint_type_group    status_group  
0           communal standpipe    communal standpipe      functional  
1           communal standpipe    communal standpipe      functional  
2  communal standpipe multiple    communal standpipe      functional  
3  communal standpipe multiple    communal standpipe  non functional  
4           communal standpipe    communal standpipe      functional

(59400, 31)

In [31]:
df_cat.columns

Index(['date_recorded', 'funder', 'installer', 'wpt_name', 'basin',
       'subvillage', 'region', 'lga', 'ward', 'public_meeting', 'recorded_by',
       'scheme_management', 'scheme_name', 'permit', 'extraction_type',
       'extraction_type_group', 'extraction_type_class', 'management',
       'management_group', 'payment', 'payment_type', 'water_quality',
       'quality_group', 'quantity', 'quantity_group', 'source', 'source_type',
       'source_class', 'waterpoint_type', 'waterpoint_type_group',
       'status_group'],
      dtype='object')

In [32]:
#let´s analyse each column to decide which one are important to consider for our model and wich are not. 

In [33]:
#checking unique values and the exact value and information for each one.
df_cat.nunique()
for x in df_cat:
    df_cat[x].value_counts()

date_recorded              356
funder                    1897
installer                 2145
wpt_name                 37400
basin                        9
subvillage               19287
region                      21
lga                        125
ward                      2092
public_meeting               2
recorded_by                  1
scheme_management           12
scheme_name               2696
permit                       2
extraction_type             18
extraction_type_group       13
extraction_type_class        7
management                  12
management_group             5
payment                      7
payment_type                 7
water_quality                8
quality_group                6
quantity                     5
quantity_group               5
source                      10
source_type                  7
source_class                 3
waterpoint_type              7
waterpoint_type_group        6
status_group                 3
dtype: int64

2011-03-15    572
2011-03-17    558
2013-02-03    546
2011-03-14    520
2011-03-16    513
             ... 
2004-03-06      1
2011-09-09      1
2011-09-18      1
2002-10-14      1
2004-07-01      1
Name: date_recorded, Length: 356, dtype: int64

Government Of Tanzania    9084
Danida                    3114
Hesawa                    2202
Rwssp                     1374
World Bank                1349
                          ... 
Abdul                        1
Samsoni                      1
Africaone Ltd                1
Bathlomew Vicent             1
Jeshi Lawokovu               1
Name: funder, Length: 1897, dtype: int64

DWE                        17402
Government                  1825
RWE                         1206
Commu                       1060
DANIDA                      1050
                           ...  
NMDC INDIA                     1
FRESH WATER PLC ENGLAND        1
Ahmad                          1
Moshono ADP                    1
NZILA                          1
Name: installer, Length: 2145, dtype: int64

none                 3563
Shuleni              1748
Zahanati              830
Msikitini             535
Kanisani              323
                     ... 
Mwabaseki A             1
Kwa Morbert             1
Kwa Olay Kashesha       1
Kwa Chemdo              1
Paloko                  1
Name: wpt_name, Length: 37400, dtype: int64

Lake Victoria              10248
Pangani                     8940
Rufiji                      7976
Internal                    7785
Lake Tanganyika             6432
Wami / Ruvu                 5987
Lake Nyasa                  5085
Ruvuma / Southern Coast     4493
Lake Rukwa                  2454
Name: basin, dtype: int64

Madukani           508
Shuleni            506
Majengo            502
Kati               373
Mtakuja            262
                  ... 
Tigiri               1
Kafeni               1
Nationanal Park      1
Ngwangilo            1
Janda Nbele          1
Name: subvillage, Length: 19287, dtype: int64

Iringa           5294
Shinyanga        4982
Mbeya            4639
Kilimanjaro      4379
Morogoro         4006
Arusha           3350
Kagera           3316
Mwanza           3102
Kigoma           2816
Ruvuma           2640
Pwani            2635
Tanga            2547
Dodoma           2201
Singida          2093
Mara             1969
Tabora           1959
Rukwa            1808
Mtwara           1730
Manyara          1583
Lindi            1546
Dar es Salaam     805
Name: region, dtype: int64

Njombe          2503
Arusha Rural    1252
Moshi Rural     1251
Bariadi         1177
Rungwe          1106
                ... 
Moshi Urban       79
Kigoma Urban      71
Arusha Urban      63
Lindi Urban       21
Nyamagana          1
Name: lga, Length: 125, dtype: int64

Igosi          307
Imalinyi       252
Siha Kati      232
Mdandu         231
Nduruma        217
              ... 
Machinjioni      1
Kirongo          1
Ikweha           1
Simbay           1
Burungura        1
Name: ward, Length: 2092, dtype: int64

True     51011
False     5055
Name: public_meeting, dtype: int64

GeoData Consultants Ltd    59400
Name: recorded_by, dtype: int64

VWC                 36793
WUG                  5206
Water authority      3153
WUA                  2883
Water Board          2748
Parastatal           1680
Private operator     1063
Company              1061
Other                 766
SWC                    97
Trust                  72
None                    1
Name: scheme_management, dtype: int64

K                           682
None                        644
Borehole                    546
Chalinze wate               405
M                           400
                           ... 
Mradi wa maji wa kipampa      1
Kisa                          1
BL Nshere Juu                 1
TM part two                   1
Ukuu                          1
Name: scheme_name, Length: 2696, dtype: int64

True     38852
False    17492
Name: permit, dtype: int64

gravity                      26780
nira/tanira                   8154
other                         6430
submersible                   4764
swn 80                        3670
mono                          2865
india mark ii                 2400
afridev                       1770
ksb                           1415
other - rope pump              451
other - swn 81                 229
windmill                       117
india mark iii                  98
cemo                            90
other - play pump               85
walimi                          48
climax                          32
other - mkulima/shinyanga        2
Name: extraction_type, dtype: int64

gravity            26780
nira/tanira         8154
other               6430
submersible         6179
swn 80              3670
mono                2865
india mark ii       2400
afridev             1770
rope pump            451
other handpump       364
other motorpump      122
wind-powered         117
india mark iii        98
Name: extraction_type_group, dtype: int64

gravity         26780
handpump        16456
other            6430
submersible      6179
motorpump        2987
rope pump         451
wind-powered      117
Name: extraction_type_class, dtype: int64

vwc                 40507
wug                  6515
water board          2933
wua                  2535
private operator     1971
parastatal           1768
water authority       904
other                 844
company               685
unknown               561
other - school         99
trust                  78
Name: management, dtype: int64

user-group    52490
commercial     3638
parastatal     1768
other           943
unknown         561
Name: management_group, dtype: int64

never pay                25348
pay per bucket            8985
pay monthly               8300
unknown                   8157
pay when scheme fails     3914
pay annually              3642
other                     1054
Name: payment, dtype: int64

never pay     25348
per bucket     8985
monthly        8300
unknown        8157
on failure     3914
annually       3642
other          1054
Name: payment_type, dtype: int64

soft                  50818
salty                  4856
unknown                1876
milky                   804
coloured                490
salty abandoned         339
fluoride                200
fluoride abandoned       17
Name: water_quality, dtype: int64

good        50818
salty        5195
unknown      1876
milky         804
colored       490
fluoride      217
Name: quality_group, dtype: int64

enough          33186
insufficient    15129
dry              6246
seasonal         4050
unknown           789
Name: quantity, dtype: int64

enough          33186
insufficient    15129
dry              6246
seasonal         4050
unknown           789
Name: quantity_group, dtype: int64

spring                  17021
shallow well            16824
machine dbh             11075
river                    9612
rainwater harvesting     2295
hand dtw                  874
lake                      765
dam                       656
other                     212
unknown                    66
Name: source, dtype: int64

spring                  17021
shallow well            16824
borehole                11949
river/lake              10377
rainwater harvesting     2295
dam                       656
other                     278
Name: source_type, dtype: int64

groundwater    45794
surface        13328
unknown          278
Name: source_class, dtype: int64

communal standpipe             28522
hand pump                      17488
other                           6380
communal standpipe multiple     6103
improved spring                  784
cattle trough                    116
dam                                7
Name: waterpoint_type, dtype: int64

communal standpipe    34625
hand pump             17488
other                  6380
improved spring         784
cattle trough           116
dam                       7
Name: waterpoint_type_group, dtype: int64

functional                 32259
non functional             22824
functional needs repair     4317
Name: status_group, dtype: int64

**date_recorded** -DROP

The date the row was entered.Should not influence the functional state of the water pump. We drop it from our data set.

**funder** - DROP

Who funded the well

#Nominal categorical data with 1897 unique values.  Who founded the well should not influence the functional state of the water pump. We drop it from our data set.

**instaler** -DROP

Who install the waterpump.

Nominal categorial data with 2145 different values. Who is the installer of  the well should not influence the functional state of the water pump. We drop it from our data set.


**wpt_name**--DROP

The name of the waterpoint

Nominal categorial data with 37400 different values. Should not influence the functional state of the water pump. We drop it from our data set.

**basin** -- KEEP IT ( regional feature)

Geographic water basin

9 unique values. 


In [34]:
# plt.figure(figsize= (12, 7), dpi= 80)
# ax = sns.countplot(x="status_group", hue='basin', data=df,order=df.sort_values('basin').status_group)
# # sns.boxplot(x='Sales Channel', y='Response', data'=data)
# plt.show()

**COLUMNS RELATED TO GEOGRAPHIC LOCATION** - 

subvillage               19287---DROP

region                      21----DROP

lga                        125---DROP

ward                      2092---DROP

In [35]:
#Maybe I could select one and try to code them between rural, urban and uknown areas??If I have time....

In [36]:
# df_cat[['subvillage', 'region', 'lga', 'ward']].nunique()

In [37]:
# We will analyses geographical situation using graphs...but not inlclude this data in our model.  
#According to the unique values of each column we will consider region and lga to group the values, however in our numerical
#data we have columns for region and district already codificated and also lat and long. 
#For this reaseon we will drop these location related categorical features.

**'PUBLIC MEETING** -- KEEP IT

True/False values. Boolean. What does it means?? Maybe it is related to the type of concesion?

In [38]:
# see if we can drop it...If I have time

**RECORDED_BY** --DROP

Same value. Not interesting for our analysis

**FEATURES RELATED TO MANAGEMENT**--REDUNDANT. SELECT ONE FOR THE MODEL --- We only keep MANAGEMENT_GROUP

scheme_management           12 --DROP

scheme_name               2696  --DROP

management                  12 -- DROP

management_group             5 ---KEEP

In [39]:
# df_cat[['scheme_management', 'scheme_name','management','management_group']].nunique()

In [40]:
# df_cat[['scheme_management', 'scheme_name','management','management_group']].isnull().sum()

In [41]:
#scheme_management,scheme_name, management has the same info about who is in charge of the management. 
#There are a lot of Nan Values in scheme_management,scheme_name so we delete this columns and keep management.
# It seems that  management and management_group provide also the same info. Let´s check it. 

In [42]:
# df_cat[['management','management_group']].value_counts()

In [43]:
# It seems that management_group groups the data included in management. We will only  keep the column of 'Management_group'for model.

**'permit'** --If the waterpoint is permitted

In [44]:
# df_cat['permit'].value_counts()
#Boolean value. We keep it for our analysis

**FEATURES RELATED TO EXTRACTION TYPE** ----REDUNDANT. SELECT ONE FOR THE MODEL

extraction_type             18 --DROP

extraction_type_group       13 --DROP

extraction_type_class        7 --KEEP


In [45]:
# df_cat[['extraction_type','extraction_type_group', 'extraction_type_class']].value_counts()

In [46]:
# Extraction_type_clase groups the data in the other two colums. Thus, we only keep this column.

**FEATURES RELATED TO PAYMENT**
payment                      7 --drop

payment_type                 7 --keep

In [47]:
# df_cat[['payment','payment_type']].value_counts()
#we will drop payment and keep 'payment type'Maybe for our model it is only interesting if it is payed or not??

**FEATURES RELATED TO WATER QUALITY/QUANTITY**

water_quality                8  ---DROP

quality_group                6 -- KEEP
 
quantity                     5  -- DROP

quantity_group               5  -KEEP

In [48]:
# df_cat[['water_quality','quality_group', 'quantity', 'quantity_group']].value_counts()
# # water quality and quality group have same info. We keep quality group column.
# # quantiy and quantity group have same info. We keep quantity_group columns

**FEATURES RELATED TO SOURCE** 

source                      10 ---DROP

source_type                  7 ---DROP

source_class                 3 --KEEP


In [49]:
# df_cat[['source', 'source_type','source_class']].value_counts()
#'source', 'source_type','source_class' have the same info. Source_class is the one which group the info. We only keep this column for our model. 

**FEATURES RELATED TO WP TYPE** 

waterpoint_type              7 --DROP

waterpoint_type_group        6 -- KEEP

In [50]:
 df_cat[[ 'waterpoint_type', 'waterpoint_type_group']].value_counts()
# #Both columns provides same info. We keep wp type group.

waterpoint_type              waterpoint_type_group
communal standpipe           communal standpipe       28522
hand pump                    hand pump                17488
other                        other                     6380
communal standpipe multiple  communal standpipe        6103
improved spring              improved spring            784
cattle trough                cattle trough              116
dam                          dam                          7
dtype: int64

**According to the above conclusion we drop selected columns from our df_cat**

In [51]:
 df_cat=df_cat.drop(['date_recorded','funder', 'installer', 'wpt_name','recorded_by','scheme_management', 'scheme_name','extraction_type',
'extraction_type_group', 'management','payment', 'water_quality','quantity','source', 'source_type',
'waterpoint_type_group','subvillage', 'region', 'lga', 'ward'],axis=1)

In [52]:
df_cat.shape
df_cat

(59400, 11)

basin public_meeting permit extraction_type_class  \
0                   Lake Nyasa           True  False               gravity   
1                Lake Victoria            NaN   True               gravity   
2                      Pangani           True   True               gravity   
3      Ruvuma / Southern Coast           True   True           submersible   
4                Lake Victoria           True   True               gravity   
...                        ...            ...    ...                   ...   
59395                  Pangani           True   True               gravity   
59396                   Rufiji           True   True               gravity   
59397                   Rufiji           True  False              handpump   
59398                   Rufiji           True   True              handpump   
59399              Wami / Ruvu           True   True              handpump   

      management_group payment_type quality_group quantity_group source_class  \
0           user-group     annually          good         enough  groundwater   
1           user-group    never pay          good   insufficient      surface   
2           user-group   per bucket          good         enough      surface   
3           user-group    never pay          good            dry  groundwater   
4                other    never pay          good       seasonal      surface   
...                ...          ...           ...            ...          ...   
59395       user-group   per bucket          good         enough  groundwater   
59396       user-group     annually          good         enough      surface   
59397       user-group      monthly      fluoride         enough  groundwater   
59398       user-group    never pay          good   insufficient  groundwater   
59399       user-group   on failure         salty         enough  groundwater   

                   waterpoint_type    status_group  
0               communal standpipe      functional  
1               communal standpipe      functional  
2      communal standpipe multiple      functional  
3      communal standpipe multiple  non functional  
4               communal standpipe      functional  
...                            ...             ...  
59395           communal standpipe      functional  
59396           communal standpipe      functional  
59397                    hand pump      functional  
59398                    hand pump      functional  
59399                    hand pump      functional  

[59400 rows x 11 columns]

In [53]:
#Dealing with Nan values in our categorical data set.

In [54]:
df_cat.isnull().sum() #HOW TO DEAL WITH NANS IN BOOLEANS. 

basin                       0
public_meeting           3334
permit                   3056
extraction_type_class       0
management_group            0
payment_type                0
quality_group               0
quantity_group              0
source_class                0
waterpoint_type             0
status_group                0
dtype: int64

In [55]:
#Dealing with Nan values 
df_cat['public_meeting'] = df_cat['public_meeting'].fillna('Unknown')
df_cat['permit'] = df_cat['permit'].fillna('Unknown')
#drop

In [56]:
# df_cat[['public_meeting']].value_counts()
# df_cat[['permit']].value_counts()

In [57]:
# def clean_public_meeting(x):
#     if 'True'in x:
#         return 'Yes'
#     elif 'False' in x:
#         return 'No'
#     else: 
#         return 'Unknown'

In [58]:
# df_cat['public_meeting']= df_cat['public_meeting'].apply (clean_public_meeting)
# df_cat['public_meeting'].nunique

In [59]:
# df_cat['permit']=df_cat['permit'].astype(str)
# df_cat['public_meeting']=df_cat['public_meeting'].astype(str)
# df_cat['permit'].dtypes
# df_cat['public_meeting'].dtypes


In [60]:
#doing some plots

In [61]:
# sns.countplot(x='status_group', data=df)
# plt.show()

In [62]:
# ax = sns.countplot(x="status_group", hue='permit', data=df)
# # sns.boxplot(x='Sales Channel', y='Response', data'=data)
# plt.show()

In [63]:
# ax = sns.countplot(x="status_group", hue='public_meeting', data=df)
# plt.show()

**Checking the correlation between categorical columns**

In [64]:
# Initally I think there colud be a dependency between  extraction_type_class and water_point_class.
#Let´s check it. NOT DOING FOR THE MOMENT

In [65]:
# data_crosstab = pd.crosstab(df_cat['extraction_type_class'], df_cat['waterpoint_type'], margins = False) # marging= False I dont want the bottom row and column that sum up values.
# data_crosstab

In [66]:
# st.chi2_contingency(data_crosstab)
# #stats
# # p-value ---o.o...it seems that there is dependency between the variables. 
# #

**Checking redundant in the values for each column**

In [67]:
def clean_waterpointype(x):
    if 'communal standpipe multiple'in x:
        return 'communal standpipe'
    elif 'communal standpipe' in x:
        return 'communal standpipe'
    else: 
        return x

In [68]:
df_cat['waterpoint_type']= df_cat['waterpoint_type'].apply (clean_waterpointype)
df_cat['waterpoint_type'].value_counts()

communal standpipe    34625
hand pump             17488
other                  6380
improved spring         784
cattle trough           116
dam                       7
Name: waterpoint_type, dtype: int64

In [69]:
df_cat['payment_type'].value_counts()

never pay     25348
per bucket     8985
monthly        8300
unknown        8157
on failure     3914
annually       3642
other          1054
Name: payment_type, dtype: int64

In [70]:
def clean_payment_status(x):
    if 'never pay'in x:
        return 'No payment'
    elif 'per bucket'or'monthly'or'on failure'or 'annually'or'other' in x:
        return 'Payment'
    elif 'unknown' in x:
        return 'unknown'
    else: 
        return x

In [71]:
df_cat['payment_type']= df_cat['payment_type'].apply (clean_payment_status)
df_cat['payment_type'].value_counts()

Payment       34052
No payment    25348
Name: payment_type, dtype: int64

#### MY FINAL DATA SET WITH THE VARIABLES SELECTED BEFORE DOING THE TEST/TRAIN SPLIT

In [72]:
df_final= pd.concat([df_num, df_cat], axis=1) 
df_final.head()

population  years_since_construction                    basin  \
0         109                        22               Lake Nyasa   
1         280                        11            Lake Victoria   
2         250                        12                  Pangani   
3          58                        35  Ruvuma / Southern Coast   
4           0                         0            Lake Victoria   

  public_meeting permit extraction_type_class management_group payment_type  \
0           True  False               gravity       user-group      Payment   
1        Unknown   True               gravity       user-group   No payment   
2           True   True               gravity       user-group      Payment   
3           True   True           submersible       user-group   No payment   
4           True   True               gravity            other   No payment   

  quality_group quantity_group source_class     waterpoint_type  \
0          good         enough  groundwater  communal standpipe   
1          good   insufficient      surface  communal standpipe   
2          good         enough      surface  communal standpipe   
3          good            dry  groundwater  communal standpipe   
4          good       seasonal      surface  communal standpipe   

     status_group  
0      functional  
1      functional  
2      functional  
3  non functional  
4      functional

### SPLIT X/Y AND TEST/TRAIN

In [73]:
y=df_final['status_group']
X=df_final.drop(['status_group'],axis=1)
from sklearn.model_selection import train_test_split
#train test
X_train,X_test,y_train,y_test = train_test_split(X,y, test_size =0.2, random_state=100)
X_train.shape
X_test.shape

(47520, 12)

(11880, 12)

### PROCESSING THE TRAIN DATA

##### 1.Processing NUMERICAL features

In [74]:
X_train_num= X_train.select_dtypes(np.number)
X_train_num

population  years_since_construction
33095         345                        10
57141         350                        35
47038           0                         0
51854           0                         0
32657           0                         0
...           ...                       ...
12119           0                         0
57191          30                        11
14147         350                        37
56088           0                         0
38408           0                         0

[47520 rows x 2 columns]

**Dealing with 0 values (=Nan)**

In [75]:
#we are going to use the median for the population POPULATION when is 0
#(NOTE: The best approach would to estimate the population according to the others geographical features.i.e same regio, village, latitude, longitude )
X_train_num['population'] =  np.where(X_train_num['population']== 0 ,X_train_num['population'].median(),X_train_num['population'])

C:\Users\Elena\anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [76]:
#we are going to use the mean for the YEARS_SINCE_CONSTRUCTION when is 0
X_train_num['years_since_construction'] =  np.where(X_train_num['years_since_construction']== 0 ,X_train_num['years_since_construction'].mean(),X_train_num['years_since_construction'])


C:\Users\Elena\anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [77]:
from sklearn.preprocessing import StandardScaler
transformer = StandardScaler().fit(X_train_num) #this line teaches standardscaler where points are, subtracting mean to it, and dividing by standard deviation
X_train_standardized = pd.DataFrame(transformer.transform(X_train_num),columns = X_train_num.columns)
X_train_standardized

population  years_since_construction
0        0.332435                 -1.036914
1        0.343082                  1.269113
2       -0.349005                 -0.505571
3       -0.349005                 -0.505571
4       -0.349005                 -0.505571
...           ...                       ...
47515   -0.349005                 -0.505571
47516   -0.338357                 -0.944673
47517    0.343082                  1.453595
47518   -0.349005                 -0.505571
47519   -0.349005                 -0.505571

[47520 rows x 2 columns]

### 2.Processing CATEGORICAL features

In [78]:
X_train_cat= X_train.select_dtypes(np.object)
X_train_cat

basin public_meeting   permit extraction_type_class  \
33095                 Internal           True     True              handpump   
57141  Ruvuma / Southern Coast           True     True               gravity   
47038               Lake Nyasa           True  Unknown               gravity   
51854               Lake Nyasa           True    False               gravity   
32657          Lake Tanganyika           True     True              handpump   
...                        ...            ...      ...                   ...   
12119              Wami / Ruvu           True     True             motorpump   
57191              Wami / Ruvu           True     True                 other   
14147  Ruvuma / Southern Coast           True     True               gravity   
56088          Lake Tanganyika           True     True              handpump   
38408                   Rufiji           True     True               gravity   

      management_group payment_type quality_group quantity_group source_class  \
33095       user-group      Payment          good         enough  groundwater   
57141       user-group      Payment          good         enough      surface   
47038       user-group      Payment          good   insufficient  groundwater   
51854       user-group      Payment          good   insufficient  groundwater   
32657       user-group      Payment          good         enough  groundwater   
...                ...          ...           ...            ...          ...   
12119       user-group      Payment          good   insufficient  groundwater   
57191       user-group   No payment         salty            dry  groundwater   
14147       user-group      Payment          good            dry      surface   
56088       user-group   No payment          good   insufficient  groundwater   
38408       user-group      Payment          good         enough  groundwater   

          waterpoint_type  
33095           hand pump  
57141  communal standpipe  
47038  communal standpipe  
51854  communal standpipe  
32657           hand pump  
...                   ...  
12119  communal standpipe  
57191               other  
14147  communal standpipe  
56088           hand pump  
38408  communal standpipe  

[47520 rows x 10 columns]

In [79]:
X_train_cat.columns
X_train_cat.dtypes

Index(['basin', 'public_meeting', 'permit', 'extraction_type_class',
       'management_group', 'payment_type', 'quality_group', 'quantity_group',
       'source_class', 'waterpoint_type'],
      dtype='object')

basin                    object
public_meeting           object
permit                   object
extraction_type_class    object
management_group         object
payment_type             object
quality_group            object
quantity_group           object
source_class             object
waterpoint_type          object
dtype: object

In [80]:
# X_train_cat_onehot = X_train_cat[['basin','extraction_type_class',
#        'management_group', 'payment_type', 'quality_group', 'quantity_group',
#        'source_class', 'waterpoint_type']]
# X_train_cat_onehot.shape

In [81]:
# from sklearn.preprocessing import OneHotEncoder
# encoder = OneHotEncoder(handle_unknown='error', drop='first').fit(X_train_cat) #drop=first, drops one of the values so we don't have redundant
# categoricals_train_encoded = encoder.transform(X_train_cat).toarray()


In [82]:
# # enc = [columname for sublist in encoder.categories_ for columname in sublist]
# X_train_onehot = pd.DataFrame(categoricals_train_encoded,columns=encoder.categories_)

In [83]:
X_cat_encoded=pd.get_dummies(X_train_cat, drop_first=True)
X_cat_encoded

basin_Lake Nyasa  basin_Lake Rukwa  basin_Lake Tanganyika  \
33095                 0                 0                      0   
57141                 0                 0                      0   
47038                 1                 0                      0   
51854                 1                 0                      0   
32657                 0                 0                      1   
...                 ...               ...                    ...   
12119                 0                 0                      0   
57191                 0                 0                      0   
14147                 0                 0                      0   
56088                 0                 0                      1   
38408                 0                 0                      0   

       basin_Lake Victoria  basin_Pangani  basin_Rufiji  \
33095                    0              0             0   
57141                    0              0             0   
47038                    0              0             0   
51854                    0              0             0   
32657                    0              0             0   
...                    ...            ...           ...   
12119                    0              0             0   
57191                    0              0             0   
14147                    0              0             0   
56088                    0              0             0   
38408                    0              0             1   

       basin_Ruvuma / Southern Coast  basin_Wami / Ruvu  public_meeting_True  \
33095                              0                  0                    1   
57141                              1                  0                    1   
47038                              0                  0                    1   
51854                              0                  0                    1   
32657                              0                  0                    1   
...                              ...                ...                  ...   
12119                              0                  1                    1   
57191                              0                  1                    1   
14147                              1                  0                    1   
56088                              0                  0                    1   
38408                              0                  0                    1   

       public_meeting_Unknown  permit_True  permit_Unknown  \
33095                       0            1               0   
57141                       0            1               0   
47038                       0            0               1   
51854                       0            0               0   
32657                       0            1               0   
...                       ...          ...             ...   
12119                       0            1               0   
57191                       0            1               0   
14147                       0            1               0   
56088                       0            1               0   
38408                       0            1               0   

       extraction_type_class_handpump  extraction_type_class_motorpump  \
33095                               1                                0   
57141                               0                                0   
47038                               0                                0   
51854                               0                                0   
32657                               1                                0   
...                               ...                              ...   
12119                               0                                1   
57191                               0                                0   
14147                               0                                0   
56088                              

**Getting final data train set**

In [84]:
X_train_standardized.reset_index(drop=True, inplace=True)
X_cat_encoded.reset_index(drop=True, inplace=True)

In [85]:

X_train_preproc = pd.concat([X_train_standardized,X_cat_encoded], axis= 1)
X_train_preproc

population  years_since_construction  basin_Lake Nyasa  \
0        0.332435                 -1.036914                 0   
1        0.343082                  1.269113                 0   
2       -0.349005                 -0.505571                 1   
3       -0.349005                 -0.505571                 1   
4       -0.349005                 -0.505571                 0   
...           ...                       ...               ...   
47515   -0.349005                 -0.505571                 0   
47516   -0.338357                 -0.944673                 0   
47517    0.343082                  1.453595                 0   
47518   -0.349005                 -0.505571                 0   
47519   -0.349005                 -0.505571                 0   

       basin_Lake Rukwa  basin_Lake Tanganyika  basin_Lake Victoria  \
0                     0                      0                    0   
1                     0                      0                    0   
2                     0                      0                    0   
3                     0                      0                    0   
4                     0                      1                    0   
...                 ...                    ...                  ...   
47515                 0                      0                    0   
47516                 0                      0                    0   
47517                 0                      0                    0   
47518                 0                      1                    0   
47519                 0                      0                    0   

       basin_Pangani  basin_Rufiji  basin_Ruvuma / Southern Coast  \
0                  0             0                              0   
1                  0             0                              1   
2                  0             0                              0   
3                  0             0                              0   
4                  0             0                              0   
...              ...           ...                            ...   
47515              0             0                              0   
47516              0             0                              0   
47517              0             0                              1   
47518              0             0                              0   
47519              0             1                              0   

       basin_Wami / Ruvu  public_meeting_True  public_meeting_Unknown  \
0                      0                    1                       0   
1                      0                    1                       0   
2                      0                    1                       0   
3                      0                    1                       0   
4                      0                    1                       0   
...                  ...                  ...                     ...   
47515                  1                    1                       0   
47516                  1                    1                       0   
47517                  0                    1                       0   
47518                  0                    1                       0   
47519                  0                    1                       0   

       permit_True  permit_Unknown  extraction_type_class_handpump  \
0                1               0                               1   
1                1               0                               0   
2                0               1                               0   
3                0               0                               0   
4                1               0                               1   
...            ...             ...                             ...   
47515            1               0                               0   
47516            1               0                               0   
47517            1               0   

In [86]:
# from sklearn.linear_model import LogisticRegression
# classification = LogisticRegression(random_state=0, solver='lbfgs',
#                   multi_class='ovr').fit(X_train_preproc, y_train)

### MODELING AND VALIDATION

#### PREPROCESSIN TEST SET

**Numericals**

In [87]:
X_test_num= X_test.select_dtypes(np.number)
X_test_num

population  years_since_construction
58203         230                        18
21703           1                        54
11362           1                        13
55745           1                        37
25523           0                         0
...           ...                       ...
13985         250                        49
40398           0                        26
48649         800                        17
499            50                        12
34783           0                         0

[11880 rows x 2 columns]

In [88]:
X_test_num['population'] =  np.where(X_test_num['population']== 0 ,X_test_num['population'].median(),X_test_num['population'])

C:\Users\Elena\anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [89]:
X_test_num['years_since_construction'] =  np.where(X_test_num['years_since_construction']== 0 ,X_test_num['years_since_construction'].mean(),X_test_num['years_since_construction'])

C:\Users\Elena\anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [90]:
from sklearn.preprocessing import StandardScaler
# transformer = StandardScaler().fit(X_test_num) #this line teaches standardscaler where points are, subtracting mean to it, and dividing by standard deviation
X_test_standardized = pd.DataFrame(transformer.transform(X_test_num),columns = X_test_num.columns)
X_test_standardized


population  years_since_construction
0        0.087542                 -0.298986
1       -0.400113                  3.021693
2       -0.400113                 -0.760191
3       -0.400113                  1.453595
4       -0.353264                 -0.508772
...           ...                       ...
11875    0.130132                  2.560488
11876   -0.353264                  0.438943
11877    1.301356                 -0.391227
11878   -0.295767                 -0.852432
11879   -0.353264                 -0.508772

[11880 rows x 2 columns]

**Categoricals**

In [91]:
X_test_cat= X_test.select_dtypes(np.object)
X_test_cat

basin public_meeting permit extraction_type_class  \
58203         Internal           True   True               gravity   
21703          Pangani           True   True               gravity   
11362  Lake Tanganyika          False  False              handpump   
55745  Lake Tanganyika          False   True              handpump   
25523         Internal        Unknown   True              handpump   
...                ...            ...    ...                   ...   
13985  Lake Tanganyika           True   True               gravity   
40398       Lake Nyasa           True  False               gravity   
48649          Pangani           True  False               gravity   
499        Wami / Ruvu           True  False           submersible   
34783    Lake Victoria           True  False               gravity   

      management_group payment_type quality_group quantity_group source_class  \
58203       user-group   No payment          good   insufficient  groundwater   
21703       user-group   No payment          good         enough  groundwater   
11362       user-group   No payment          good         enough  groundwater   
55745            other      Payment          good         enough  groundwater   
25523       user-group      Payment         salty   insufficient  groundwater   
...                ...          ...           ...            ...          ...   
13985       user-group   No payment          good   insufficient      surface   
40398       user-group      Payment          good         enough  groundwater   
48649       user-group   No payment          good   insufficient  groundwater   
499         user-group      Payment          good         enough  groundwater   
34783       user-group   No payment          good            dry  groundwater   

          waterpoint_type  
58203  communal standpipe  
21703  communal standpipe  
11362           hand pump  
55745           hand pump  
25523           hand pump  
...                   ...  
13985  communal standpipe  
40398  communal standpipe  
48649  communal standpipe  
499    communal standpipe  
34783  communal standpipe  

[11880 rows x 10 columns]

In [92]:
X_cat_test_encoded=pd.get_dummies(X_test_cat, drop_first=True)
X_cat_test_encoded

basin_Lake Nyasa  basin_Lake Rukwa  basin_Lake Tanganyika  \
58203                 0                 0                      0   
21703                 0                 0                      0   
11362                 0                 0                      1   
55745                 0                 0                      1   
25523                 0                 0                      0   
...                 ...               ...                    ...   
13985                 0                 0                      1   
40398                 1                 0                      0   
48649                 0                 0                      0   
499                   0                 0                      0   
34783                 0                 0                      0   

       basin_Lake Victoria  basin_Pangani  basin_Rufiji  \
58203                    0              0             0   
21703                    0              1             0   
11362                    0              0             0   
55745                    0              0             0   
25523                    0              0             0   
...                    ...            ...           ...   
13985                    0              0             0   
40398                    0              0             0   
48649                    0              1             0   
499                      0              0             0   
34783                    1              0             0   

       basin_Ruvuma / Southern Coast  basin_Wami / Ruvu  public_meeting_True  \
58203                              0                  0                    1   
21703                              0                  0                    1   
11362                              0                  0                    0   
55745                              0                  0                    0   
25523                              0                  0                    0   
...                              ...                ...                  ...   
13985                              0                  0                    1   
40398                              0                  0                    1   
48649                              0                  0                    1   
499                                0                  1                    1   
34783                              0                  0                    1   

       public_meeting_Unknown  permit_True  permit_Unknown  \
58203                       0            1               0   
21703                       0            1               0   
11362                       0            0               0   
55745                       0            1               0   
25523                       1            1               0   
...                       ...          ...             ...   
13985                       0            1               0   
40398                       0            0               0   
48649                       0            0               0   
499                         0            0               0   
34783                       0            0               0   

       extraction_type_class_handpump  extraction_type_class_motorpump  \
58203                               0                                0   
21703                               0                                0   
11362                               1                                0   
55745                               1                                0   
25523                               1                                0   
...                               ...                              ...   
13985                               0                                0   
40398                               0                                0   
48649                               0                                0   
499                                

**Getting final data test set**

In [93]:
X_test_standardized.reset_index(drop=True, inplace=True)
X_cat_test_encoded.reset_index(drop=True, inplace=True)

In [94]:
X_test_preproc = pd.concat([X_test_standardized,X_cat_test_encoded], axis= 1)
X_test_preproc

population  years_since_construction  basin_Lake Nyasa  \
0        0.087542                 -0.298986                 0   
1       -0.400113                  3.021693                 0   
2       -0.400113                 -0.760191                 0   
3       -0.400113                  1.453595                 0   
4       -0.353264                 -0.508772                 0   
...           ...                       ...               ...   
11875    0.130132                  2.560488                 0   
11876   -0.353264                  0.438943                 1   
11877    1.301356                 -0.391227                 0   
11878   -0.295767                 -0.852432                 0   
11879   -0.353264                 -0.508772                 0   

       basin_Lake Rukwa  basin_Lake Tanganyika  basin_Lake Victoria  \
0                     0                      0                    0   
1                     0                      0                    0   
2                     0                      1                    0   
3                     0                      1                    0   
4                     0                      0                    0   
...                 ...                    ...                  ...   
11875                 0                      1                    0   
11876                 0                      0                    0   
11877                 0                      0                    0   
11878                 0                      0                    0   
11879                 0                      0                    1   

       basin_Pangani  basin_Rufiji  basin_Ruvuma / Southern Coast  \
0                  0             0                              0   
1                  1             0                              0   
2                  0             0                              0   
3                  0             0                              0   
4                  0             0                              0   
...              ...           ...                            ...   
11875              0             0                              0   
11876              0             0                              0   
11877              1             0                              0   
11878              0             0                              0   
11879              0             0                              0   

       basin_Wami / Ruvu  public_meeting_True  public_meeting_Unknown  \
0                      0                    1                       0   
1                      0                    1                       0   
2                      0                    0                       0   
3                      0                    0                       0   
4                      0                    0                       1   
...                  ...                  ...                     ...   
11875                  0                    1                       0   
11876                  0                    1                       0   
11877                  0                    1                       0   
11878                  1                    1                       0   
11879                  0                    1                       0   

       permit_True  permit_Unknown  extraction_type_class_handpump  \
0                1               0                               0   
1                1               0                               0   
2                0               0                               1   
3                1               0                               1   
4                1               0                               1   
...            ...             ...                             ...   
11875            1               0                               0   
11876            0               0                               0   
11877            0               0   

### Logistic Regression

In [95]:
from sklearn.linear_model import LogisticRegression
classificationLR = LogisticRegression(random_state=0, solver='lbfgs',
                  multi_class='ovr').fit(X_train_preproc, y_train)

C:\Users\Elena\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
C:\Users\Elena\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-reg

In [96]:
predictionsLR = classificationLR.predict(X_test_preproc)
classificationLR.score(X_test_preproc, y_test)

0.7186868686868687

In [110]:
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score

In [117]:
print("precision: ",precision_score(y_test,predictionsLR,average='macro' ))
print("recall: ",recall_score(y_test,predictionsLR,average='macro' ))
print("f1: ",f1_score(y_test,predictionsLR,average='macro' ))

C:\Users\Elena\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


precision:  0.4926157343547355
recall:  0.4987291601909643
f1:  0.48569042583552857


In [97]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, predictionsLR)

array([[5852,    0,  569],
       [ 706,    0,  160],
       [1907,    0, 2686]], dtype=int64)

In [98]:
#              Predicted Labels
#              | A  | B
# -----------------------
# True label A |    |
#            ------------
#            B |    |

**Importances features**

In [99]:
importances = pd.DataFrame(data={
    'Attribute': X_train_preproc.columns,
    'Importance': classificationLR.coef_[0]
})
importances = importances.sort_values(by='Importance', ascending=False)
importances.head(10)

Attribute  Importance
30           quantity_group_enough    4.140924
32         quantity_group_seasonal    3.901874
31     quantity_group_insufficient    3.639778
33          quantity_group_unknown    2.999929
27             quality_group_milky    1.248589
35            source_class_unknown    1.045564
25          quality_group_fluoride    0.749912
24            payment_type_Payment    0.631162
20          management_group_other    0.619762
14  extraction_type_class_handpump    0.581723

In [100]:
# plt.bar(x=importances['Attribute'], height=importances['Importance'], color='#087E8B')
# plt.title('Feature importances obtained from coefficients', size=20)
# plt.xticks(rotation='vertical')
# plt.xlabel('Feature importance')
# plt.ylabel('Features')
# # plt.show()

### K-Nearest Neighbors

In [101]:
from sklearn.neighbors import KNeighborsClassifier
classificationKNN = KNeighborsClassifier(n_neighbors=3)    
classificationKNN.fit(X_train_preproc, y_train)

KNeighborsClassifier(n_neighbors=3)

In [102]:
predictionsKNN = classificationKNN.predict(X_test_preproc)
classificationKNN.score(X_test_preproc, y_test)

0.7353535353535353

In [116]:
print("precision: ",precision_score(y_test,predictionsKNN,average='macro' ))
print("recall: ",recall_score(y_test,predictionsKNN,average='macro'))
print("f1: ",f1_score(y_test,predictionsKNN,average='macro'))

precision:  0.6353795326597927
recall:  0.5845211771048555
f1:  0.6002203254227352


In [103]:
confusion_matrix(y_test, predictionsKNN)

array([[5343,  204,  874],
       [ 515,  195,  156],
       [1312,   83, 3198]], dtype=int64)

### Random Forest


In [104]:
from sklearn.ensemble import RandomForestClassifier
classificationRF= RandomForestClassifier(max_depth=25, n_estimators = 42*5, criterion = 'entropy', random_state = 0)
classificationRF.fit(X_train_preproc, y_train)  

RandomForestClassifier(criterion='entropy', max_depth=25, n_estimators=210,
                       random_state=0)

In [105]:
predictionsRF = classificationRF.predict(X_test_preproc)
classificationRF.score(X_test_preproc, y_test)

0.7636363636363637

In [118]:
print("precision: ",precision_score(y_test,predictionsRF,average='macro' ))
print("recall: ",recall_score(y_test,predictionsRF,average='macro'))
print("f1: ",f1_score(y_test,predictionsRF,average='macro'))

precision:  0.6727774768332456
recall:  0.5998870350843366
f1:  0.6189106644036212


In [106]:
confusion_matrix(y_test, predictionsRF)

array([[5583,  145,  693],
       [ 529,  182,  155],
       [1218,   68, 3307]], dtype=int64)

**Important Features**

In [107]:
model = classificationRF
model.fit(X_train_preproc, y_train)
importances = pd.DataFrame(data={
    'Attribute': X_train_preproc.columns,
    'Importance': model.feature_importances_
})
importancesRF = importances.sort_values(by='Importance', ascending=False)

RandomForestClassifier(criterion='entropy', max_depth=25, n_estimators=210,
                       random_state=0)

In [108]:
importancesRF.head(20)

Attribute  Importance
1            years_since_construction    0.190094
0                          population    0.187525
30              quantity_group_enough    0.058180
31        quantity_group_insufficient    0.045713
16        extraction_type_class_other    0.041357
40              waterpoint_type_other    0.037393
12                        permit_True    0.036790
24               payment_type_Payment    0.036015
32            quantity_group_seasonal    0.028962
34               source_class_surface    0.026508
23        management_group_user-group    0.022377
10                public_meeting_True    0.022145
5                 basin_Lake Victoria    0.019919
18  extraction_type_class_submersible    0.018690
4               basin_Lake Tanganyika    0.016154
7                        basin_Rufiji    0.013979
2                    basin_Lake Nyasa    0.013875
9                   basin_Wami / Ruvu    0.013734
6                       basin_Pangani    0.013651
26                 quality_group_good    0.013382